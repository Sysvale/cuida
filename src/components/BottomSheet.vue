<template>
	<div
		v-if="internalModelValue"
	>
		<div
			:class="toggleAnimationClass ? 'backdrop__show' : 'backdrop__hide'"
			@click="closeSideSheet"
		>
			<div
				ref="bottomSheetRef"
				:class="toggleAnimationClass ? 'bottom-sheet__show' : 'bottom-sheet__hide'"
			>
				<div class="bottom-sheet__header-border">
					<div class="bottom-sheet__header-border-line" />
				</div>
				<div class="bottom-sheet__header">
					<div class="bottom-sheet__title">
						{{ title }}
					</div>
					<div
						class="bottom-sheet__close-icon"
						@click="closeHandle"
					>
						<icon
							name="x-outline"
							height="20"
							width="20"
							color="#29333D"
						/>
					</div>
				</div>
				<div class="bottom-sheet__content">
					<!-- @slot Scoped slot usado para inserção de conteúdo customizado. -->
					<slot name="default" />
				</div>
			</div>
		</div>
	</div>
</template>

<script setup>
import Icon from './Icon.vue';
import { ref, watch } from 'vue';

const props = defineProps({
	modelValue: {
		type: Boolean,
		default: false,
		required: true,
	},

	/**
		 * Define o título do bottom sheet exibido no header
		 */
	title: {
		type: String,
		default: 'Onde encontrar o meu CNS',
	},
	/**
		* Define se o BottomSheet vai ser fechado com o click no backdrop.
		*/
	noCloseOnBackdrop: {
		type: Boolean,
		default: false,
	},
});

const emit = defineEmits([
	/**
	 * Evento emitido quando a visibilidade do bottom sheet é alterada.
	 * @event update:model-value
	 * @type {Event}
	 */
	'update:model-value',

	/**
	 * Evento emitido quando o bottom sheet é fechado.
	 * @event close
	 * @type {Event}
	 */
	'close',
]);

const internalModelValue = ref(props.modelValue);
const toggleAnimationClass = ref(true);
const bottomSheetRef = ref(null);

watch(() => props.modelValue, (newValue, oldValue) => {
	if (newValue !== oldValue) {
		if (!newValue) {
			closeHandle();
		} else {
			internalModelValue.value = newValue;
		}

		mustDisableExternalScrolls(newValue);
	}
});

function closeSideSheet(event) {
	if (props.noCloseOnBackdrop) {
		return;
	}

	if (bottomSheetRef.value && !bottomSheetRef.value.contains(event.target)) {
		closeHandle();
	}
}

function closeHandle() {
	toggleAnimationClass.value = false;

	setTimeout(() => {
		internalModelValue.value = false;
		toggleAnimationClass.value = true;
		emit('update:model-value', false);
		emit('close', true);
	}, 300);
}

function mustDisableExternalScrolls(value) {
	document.body.style.overflow = value ? 'hidden' : 'auto';
}

</script>

<style lang="scss" scoped>
@import '../assets/sass/tokens.scss';

.backdrop {
	display: flex;
	justify-content: center;
	align-items: center;
	position: fixed;
	padding: px(4);
	top: 0;
	bottom: 0;
	left: 0;
	right: 0;
	background-color: rgba(0, 0, 0, 0.45);
	z-index: 1060;

	&__show {
		@extend .backdrop;
		animation: zoom-in ease .3s;
	}

	&__hide {
		@extend .backdrop;
		animation: zoom-out ease .3s;
	}
}

.bottom-sheet {
	width: 100%;
    height: 75vh;
    border-top-left-radius: $border-radius-extra-large;
    border-top-right-radius: $border-radius-extra-large;
	position: absolute;
	bottom: 0;
	height: fit-content;
	z-index: $z-index-modal;
	background-color: $n-0;
	padding: pa(6);
	color: $n-800;

	&__header-border {
		display: flex;
		justify-content: center;
		align-items: center;
		margin: mTRBL(n2, 0, 4, 0);
	}

	&__header-border-line {
		width: 30px;
		height: 4px;
		background: $n-40;
		border-radius: $border-radius-large;
	}

	&__show {
		@extend .bottom-sheet;
		animation: slide-up ease .3s;
	}

	&__hide {
		@extend .bottom-sheet;
		animation: slide-down ease .3s;
	}

	&__header {
		display: flex;
		justify-content: space-between;
		align-items: center;
	}

	&__close-icon {
		cursor: pointer;
	}

	&__title {
		@include button-1;
		font-weight: 700;
	}

	&__content {
		margin: mt(6);
		max-height: 75vh;
		overflow-y: auto;
	}
}

@keyframes zoom-in {
	0% {
		opacity: 0;
	}
	100% {
		opacity: 1;
	}
}

@keyframes zoom-out {
	0% {
		opacity: 1;
	}
	100% {
		opacity: 0;
	}
}

@keyframes slide-up {
    0% {
        transform: translateY(100%);
    }
    100% {
        transform: translateY(0);
    }
}

@keyframes slide-down {
    0% {
        transform: translateY(0);
    }
    100% {
        transform: translateY(100%);
    }
}
</style>
