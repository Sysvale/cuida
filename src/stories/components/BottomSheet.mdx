import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import BottomSheet from '../../components/BottomSheet.vue';
import Button from '../../components/Button.vue';
import * as BottomSheetStories from './BottomSheet.stories';

<Meta of={BottomSheetStories} />

export const Template = (args) => ({
	components: { CdsBottomSheet: BottomSheet, CdsButton: Button },
	setup() {
		return { args };
	},
	data() {
		return {
			showBottomSheet: false,
		};
	},
	template: /*html*/ `
		<div style="display: flex; justify-content: center">
			<cds-button
				style="margin: 12px;"
				@click="showBottomSheet = true"
			>
				Click
			</cds-button>
			<cds-bottom-sheet
				v-bind="args"
				v-model="showBottomSheet"
			>
				Conteúdo
			</cds-bottom-sheet>
		</div>
	`,
});

# Bottom Sheet

### BottomSheets são componentes que exibem conteúdo adicional em uma camada sobreposta na parte inferior da página atual, fornecendo acesso rápido a informações ou ações complementares sem interromper o fluxo principal da interface.

***

<br />

## Quando usar:

* Para fornecer acesso rápido a opções ou controles adicionais sem obscurecer o conteúdo principal da página;
* Quando há necessidade de apresentar informações contextuais relevantes sem exigir uma mudança de contexto para o usuário;
* Para oferecer uma experiência de navegação fluida, permitindo que o usuário permaneça na mesma tela enquanto interage com o conteúdo adicional.

<br />

## Quando não usar:

* Se o conteúdo ou as ações adicionais exigirem mais espaço do que a área disponível na parte inferior da tela.
* Quando a presença de um BottomSheet interferiria na experiência do usuário ou causaria confusão ao sobrepor-se a outros elementos essenciais da interface.
* Se houver necessidade de exibir uma quantidade significativa de informações ou fluxos complexos que não possam ser adequadamente contidos na parte inferior da tela.

<br />

## Preview

<Canvas>
  <Story of={BottomSheetStories.BottomSheet} />
</Canvas>

<Controls />
