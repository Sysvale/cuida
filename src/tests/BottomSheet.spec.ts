import { describe, test, expect } from 'vitest';
import BottomSheet from '../components/BottomSheet.vue';
import { mount } from '@vue/test-utils';

describe('BottomSheet', () => {
	test('renders correctly', async () => {
		const wrapper = mount(BottomSheet, {
			props: {
				modelValue: true,
				title: 'Qual o segredo da vida?',
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});
});
