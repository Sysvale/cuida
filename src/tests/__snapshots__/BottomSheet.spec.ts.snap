// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BottomSheet > renders correctly 1`] = `
"<div data-v-4053a7e3="">
  <div data-v-4053a7e3="" class="backdrop__show">
    <div data-v-4053a7e3="" class="bottom-sheet__show">
      <div data-v-4053a7e3="" class="bottom-sheet__header-border">
        <div data-v-4053a7e3="" class="bottom-sheet__header-border-line"></div>
      </div>
      <div data-v-4053a7e3="" class="bottom-sheet__header">
        <div data-v-4053a7e3="" class="bottom-sheet__title">Qual o segredo da vida?</div>
        <div data-v-4053a7e3="" class="bottom-sheet__close-icon"><svg data-v-4053a7e3="" xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0 0 24 24" aria-labelledby="x-outline" role="presentation">
            <g fill="#29333D"></g>
          </svg></div>
      </div>
      <div data-v-4053a7e3="" class="bottom-sheet__content">
        <!-- @slot Scoped slot usado para inserção de conteúdo customizado. -->
      </div>
    </div>
  </div>
</div>"
`;
